byway/src/main/java/com/google/sps/servlets/EntityServlet.java

```diff
@@ -0,0 +1,115 @@
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.gson.Gson;
+import com.google.sps.data.Trip;
+import com.google.sps.data.UserInfo;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/entity")
+public class EntityServlet extends HttpServlet {
+  private final Gson gson = new Gson();
+  private final UserService userService = UserServiceFactory.getUserService();
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    com.google.appengine.api.users.User user = userService.getCurrentUser();
+    UserInfo userInfo = addUserEntity(user);
+    List<String> userTripIds = userInfo.getTripIds();
+    ArrayList<Trip> userTrips = new ArrayList<Trip>();
+    for (String tripId : userTripIds) {
+      try {
+        Entity tripEntity = datastore.get(KeyFactory.stringToKey(tripId));
+        Trip trip = Trip.fromEntity(tripEntity);
+        userTrips.add(trip);
+      } catch (EntityNotFoundException exception) {
+        return;
+      }
+    }
+    String json = gson.toJson(userTrips);
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Key tripKey = createTripEntity();
+    String keyString = KeyFactory.keyToString(tripKey);
+    System.out.println(keyString);
+    addTripForUser(keyString);
+    String json = gson.toJson(KeyFactory.keyToString(tripKey));
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  /*
+   * Adds new Trip entity with empty properties
+   **/
+  public Key createTripEntity() {
+    System.setProperty(
+        DatastoreServiceConfig.DATASTORE_EMPTY_LIST_SUPPORT, Boolean.TRUE.toString());
+    Entity tripEntity = new Entity(Trip.DATASTORE_ENTITY_KIND);
+    tripEntity.setProperty("start", "");
+    tripEntity.setProperty("destinations", new ArrayList<String>());
+    tripEntity.setProperty("interests", new ArrayList<String>());
+    tripEntity.setProperty("route", new ArrayList<String>());
+    datastore.put(tripEntity);
+    Key tripKey = tripEntity.getKey();
+    tripEntity.setProperty("keyString", KeyFactory.keyToString(tripKey));
```

Comment:
There's no need to redundantly store the entity's own key on that entity, and on the other hand it is error-prone to do so. For example, what if you accidentally set a different value for `keyString` later, so the entity's `Key` and its `keyString` don't actually refer to the same entity? That could easily cause very odd bugs, like if you load an entity by key, then save by `keyString`, you might find that you're overwriting a different entity or creating a new entity unexpectedly.

In general, you should prefer to have a Single Source of Truth regarding what the state of some data is. There are some occasional cases where optimizations require storing duplicate data, but that always should be done very carefully to ensure the data stays synced, and even then there's usually one side that's considered the official 'source of truth'. `keyToString` is probably not a sufficiently expensive operation to warrant saving the keyString on the entity (plus doing so actually will make this code slower, since it now takes two calls to `datastore.put` which is going to be much slower than `keyToString` or `stringToKey`.

Reply:
took out KeyString